ground/gcs/src/plugins/uavtalk/uavtalk.cpp

```diff
@@ -191,244 +182,128 @@ bool UAVTalk::objectTransaction(UAVObject *obj, quint8 type, bool allInstances)
 
-            // Determine data length
-            if (rxType == TYPE_OBJ_REQ || rxType == TYPE_ACK || rxType == TYPE_NACK) {
-                rxLength = 0;
-            } else {
-                rxLength = rxObj->getNumBytes();
-            }
+    if (ourCrc != *theirCrc) {
+        /* Since we can't trust hdr->size for sure, we should just skip
+         * forward one byte.
+         */
 
-            // Check length and determine next state
-            if (rxLength >= MAX_PAYLOAD_LENGTH) {
-                stats.rxErrors++;
-                rxState = STATE_SYNC;
-                UAVTALK_QXTLOG_DEBUG("UAVTalk: ObjID->Sync (oversize)");
-                break;
-            }
+        startOffset++;
+        stats.rxErrors++;
 
-            quint8 rxInstanceLength = (rxObj->isSingleInstance() ? 0 : 2);
-            if ((rxPacketLength + rxInstanceLength + rxLength)
-                != packetSize) { // packet error - mismatched packet size
-                stats.rxErrors++;
-                rxState = STATE_SYNC;
-                UAVTALK_QXTLOG_DEBUG("UAVTalk: ObjID->Sync (length mismatch)");
-                break;
-            }
+        return true;
+    }
 
-            if (rxObj->isSingleInstance()) { // Check if this is a single instance object (i.e. if
-                                             // the instance ID field is coming next)
-                // If there is a payload get it, otherwise receive checksum
-                if (rxLength > 0) {
-                    rxState = STATE_DATA;
-                    UAVTALK_QXTLOG_DEBUG("UAVTalk: ObjID->Data (needs data)");
-                } else {
-                    rxState = STATE_CS;
-                    UAVTALK_QXTLOG_DEBUG("UAVTalk: ObjID->Checksum");
-                }
-                rxInstId = 0;
-                rxCount = 0;
-            } else {
-                rxState = STATE_INSTID;
-                UAVTALK_QXTLOG_DEBUG("UAVTalk: ObjID->InstID");
-                rxCount = 0;
-            }
-        }
+    quint8 *payload = rxBuffer + startOffset + sizeof(*hdr);
+    unsigned int payloadBytes = hdr->size - sizeof(*hdr);
 
-        break;
+    /* At this point, we'll advance startOffset for the entire length of
+     * frame, and not touch startOffset again this function!
+     */
+    startOffset += hdr->size + 1;
 
-    case STATE_INSTID:
+    /* OK, we have a complete frame as encoded on the wire.  Time to do things
+     * with it.
+     */
 
-        // Update CRC
-        rxCS = updateCRC(rxCS, rxbyte);
+    quint8 rxType = hdr->type & TYPE_MASK;
 
-        rxTmpBuffer[rxCount++] = rxbyte;
-        if (rxCount < 2) {
-            UAVTALK_QXTLOG_DEBUG("UAVTalk: InstID->InstID");
-            break;
-        }
+    quint32 rxObjId = qFromLittleEndian(hdr->objId);
 
-        rxInstId = (qint16)qFromLittleEndian<quint16>(rxTmpBuffer);
+    UAVObject *rxObj = objMngr->getObject(rxObjId);
 
-        rxCount = 0;
+    if (rxObj == NULL) {
```

Comment:
ergh I hate Q_\*.  ;)

Will change-- in the future we may want to consider nullptr since we're C++11 anyways.